### ISSUE TITLE
print $outfh <$infh> gives Syntax Error

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#16242]
Migrated from [rt.perl.org#132434](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=132434) (status was 'rejected')

Searchable as RT132434$

**Comment by p5pRT** [comment]
### From @jimav

This is a bug report for perl from jim\.avera@&#8203;gmail\.com\,
generated with the help of perlbug 1\.40 running under perl 5\.24\.1\.




---
If print is called with $fh instead of an old\-style FILEHANDLE \(no comma 
in either case\)\,
then a Syntax Error occurs if the first data item is \<$fh>\.

    print $outfh \<$infh>    \#Syntax Error

No error occurs if $outfh or $infh is replaced by HANDLE\, of if \<$infh> 
is not the first data item\.

\#\!/usr/bin/perl
use strict; use warnings;
require File&#8203;::Temp;

open INFH\,     "\<"\, "/etc/timezone" or die;
open my $infh\, "\<"\, "/etc/timezone" or die;

open OUTFH\,     ">&STDOUT" or die;
open my $outfh\, ">&STDOUT" or die;

local $/;  \#slurp

print $outfh \<$infh> or die;          \# SYNTAX ERROR
print $outfh "fooey"\, \<$infh> or die; \# ok
print $outfh \<INFH> or die;           \# ok
print OUTFH \<$infh> or die;           \# ok
print OUTFH \<INFH> or die;            \# ok





---


---
Flags&#8203;:
     category=core
     severity=low


---
Site configuration information for perl 5\.24\.1&#8203;:

Configured by Debian Project at Sat Mar 18 17&#8203;:00&#8203;:39 UTC 2017\.

Summary of my perl5 \(revision 5 version 24 subversion 1\) configuration&#8203;:

   Platform&#8203;:
     osname=linux\, osvers=3\.16\.0\, archname=x86\_64\-linux\-gnu\-thread\-multi
     uname='linux localhost 3\.16\.0 \#1 smp debian 3\.16\.0 x86\_64 gnulinux '
     config\_args='\-Dusethreads \-Duselargefiles \-Dcc=x86\_64\-linux\-gnu\-gcc 
\-Dcpp=x86\_64\-linux\-gnu\-cpp \-Dld=x86\_64\-linux\-gnu\-gcc \-Dccflags=\-DDEBIAN 
\-Wdate\-time \-D\_FORTIFY\_SOURCE=2 \-g \-O2 
\-fdebug\-prefix\-map=/build/perl\-gaNj0t/perl\-5\.24\.1=\. 
\-fstack\-protector\-strong \-Wformat \-Werror=format\-security \-Dldflags= 
\-Wl\,\-Bsymbolic\-functions \-Wl\,\-z\,relro \-Dlddlflags=\-shared 
\-Wl\,\-Bsymbolic\-functions \-Wl\,\-z\,relro \-Dcccdlflags=\-fPIC 
\-Darchname=x86\_64\-linux\-gnu \-Dprefix=/usr \-Dprivlib=/usr/share/perl/5\.24 
\-Darchlib=/usr/lib/x86\_64\-linux\-gnu/perl/5\.24 \-Dvendorprefix=/usr 
\-Dvendorlib=/usr/share/perl5 
\-Dvendorarch=/usr/lib/x86\_64\-linux\-gnu/perl5/5\.24 
\-Dsiteprefix=/usr/local \-Dsitelib=/usr/local/share/perl/5\.24\.1 
\-Dsitearch=/usr/local/lib/x86\_64\-linux\-gnu/perl/5\.24\.1 
\-Dman1dir=/usr/share/man/man1 \-Dman3dir=/usr/share/man/man3 
\-Dsiteman1dir=/usr/local/man/man1 \-Dsiteman3dir=/usr/local/man/man3
\-Dusesitecustomize \-Duse64bitint \-Dman1ext=1 \-Dman3ext=3perl 
\-Dpager=/usr/bin/sensible\-pager \-Uafs \-Ud\_csh \-Ud\_ualarm \-Uusesfio 
\-Uusenm \-Ui\_libutil \-Uversiononly \-DDEBUGGING=\-g \-Doptimize=\-O2 \-dEs 
\-Duseshrplib \-Dlibperl=libperl\.so\.5\.24\.1'
     hint=recommended\, useposix=true\, d\_sigaction=define
     useithreads=define\, usemultiplicity=define
     use64bitint=define\, use64bitall=define\, uselongdouble=undef
     usemymalloc=n\, bincompat5005=undef
   Compiler&#8203;:
     cc='x86\_64\-linux\-gnu\-gcc'\, ccflags ='\-D\_REENTRANT \-D\_GNU\_SOURCE 
\-DDEBIAN \-fwrapv \-fno\-strict\-aliasing \-pipe \-I/usr/local/include 
\-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
     optimize='\-O2 \-g'\,
     cppflags='\-D\_REENTRANT \-D\_GNU\_SOURCE \-DDEBIAN \-fwrapv 
\-fno\-strict\-aliasing \-pipe \-I/usr/local/include'
     ccversion=''\, gccversion='6\.3\.0 20170316'\, gccosandvers=''
     intsize=4\, longsize=8\, ptrsize=8\, doublesize=8\, byteorder=12345678\, 
doublekind=3
     d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, 
longdblsize=16\, longdblkind=3
     ivtype='long'\, ivsize=8\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, 
lseeksize=8
     alignbytes=8\, prototype=define
   Linker and Libraries&#8203;:
     ld='x86\_64\-linux\-gnu\-gcc'\, ldflags =' \-fstack\-protector\-strong 
\-L/usr/local/lib'
     libpth=/usr/local/lib /usr/lib/gcc/x86\_64\-linux\-gnu/6/include\-fixed 
/usr/include/x86\_64\-linux\-gnu /usr/lib /lib/x86\_64\-linux\-gnu /lib/\.\./lib 
/usr/lib/x86\_64\-linux\-gnu /usr/lib/\.\./lib /lib
     libs=\-lgdbm \-lgdbm\_compat \-ldb \-ldl \-lm \-lpthread \-lc \-lcrypt
     perllibs=\-ldl \-lm \-lpthread \-lc \-lcrypt
     libc=libc\-2\.24\.so\, so=so\, useshrplib=true\, libperl=libperl\.so\.5\.24
     gnulibc\_version='2\.24'
   Dynamic Linking&#8203;:
     dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E'
     cccdlflags='\-fPIC'\, lddlflags='\-shared \-L/usr/local/lib 
\-fstack\-protector\-strong'

Locally applied patches&#8203;:
     DEBPKG&#8203;:debian/cpan\_definstalldirs \- Provide a sensible INSTALLDIRS 
default for modules installed from CPAN\.
     DEBPKG&#8203;:debian/db\_file\_ver \- https://bugs.debian.org/340047 Remove 
overly restrictive DB\_File version check\.
     DEBPKG&#8203;:debian/doc\_info \- Replace generic man\(1\) instructions with 
Debian\-specific information\.
     DEBPKG&#8203;:debian/enc2xs\_inc \- https://bugs.debian.org/290336 Tweak 
enc2xs to follow symlinks and ignore missing @&#8203;INC directories\.
     DEBPKG&#8203;:debian/errno\_ver \- https://bugs.debian.org/343351 Remove 
Errno version check due to upgrade problems with long\-running processes\.
     DEBPKG&#8203;:debian/libperl\_embed\_doc \- https://bugs.debian.org/186778 
Note that libperl\-dev package is required for embedded linking
     DEBPKG&#8203;:fixes/respect\_umask \- Respect umask during installation
     DEBPKG&#8203;:debian/writable\_site\_dirs \- Set umask approproately for site 
install directories
     DEBPKG&#8203;:debian/extutils\_set\_libperl\_path \- EU&#8203;:MM&#8203;: set location of 
libperl\.a under /usr/lib
     DEBPKG&#8203;:debian/no\_packlist\_perllocal \- Don't install \.packlist or 
perllocal\.pod for perl or vendor
     DEBPKG&#8203;:debian/fakeroot \- Postpone LD\_LIBRARY\_PATH evaluation to the 
binary targets\.
     DEBPKG&#8203;:debian/instmodsh\_doc \- Debian policy doesn't install 
\.packlist files for core or vendor\.
     DEBPKG&#8203;:debian/ld\_run\_path \- Remove standard libs from LD\_RUN\_PATH 
as per Debian policy\.
     DEBPKG&#8203;:debian/libnet\_config\_path \- Set location of libnet\.cfg to 
/etc/perl/Net as /usr may not be writable\.
     DEBPKG&#8203;:debian/mod\_paths \- Tweak @&#8203;INC ordering for Debian
     DEBPKG&#8203;:debian/prune\_libs \- https://bugs.debian.org/128355 Prune the 
list of libraries wanted to what we actually need\.
     DEBPKG&#8203;:fixes/net\_smtp\_docs \- \[rt\.cpan\.org \#36038\] 
https://bugs.debian.org/100195 Document the Net&#8203;::SMTP 'Port' option
     DEBPKG&#8203;:debian/perlivp \- https://bugs.debian.org/510895 Make perlivp 
skip include directories in /usr/local
     DEBPKG&#8203;:debian/deprecate\-with\-apt \- https://bugs.debian.org/747628 
Point users to Debian packages of deprecated core modules
     DEBPKG&#8203;:debian/squelch\-locale\-warnings \- 
https://bugs.debian.org/508764 Squelch locale warnings in Debian package 
maintainer scripts
     DEBPKG&#8203;:debian/skip\-upstream\-git\-tests \- Skip tests specific to the 
upstream Git repository
     DEBPKG&#8203;:debian/patchlevel \- https://bugs.debian.org/567489 List 
packaged patches for 5\.24\.1\-2ubuntu1 in patchlevel\.h
     DEBPKG&#8203;:debian/skip\-kfreebsd\-crash \- https://bugs.debian.org/628493 
\[perl \#96272\] Skip a crashing test case in t/op/threads\.t on GNU/kFreeBSD
     DEBPKG&#8203;:fixes/document\_makemaker\_ccflags \- 
https://bugs.debian.org/628522 \[rt\.cpan\.org \#68613\] Document that 
CCFLAGS should include $Config\{ccflags\}
     DEBPKG&#8203;:debian/find\_html2text \- https://bugs.debian.org/640479 
Configure CPAN&#8203;::Distribution with correct name of html2text
     DEBPKG&#8203;:debian/perl5db\-x\-terminal\-emulator\.patch \- 
https://bugs.debian.org/668490 Invoke x\-terminal\-emulator rather than 
xterm in perl5db\.pl
     DEBPKG&#8203;:debian/cpan\-missing\-site\-dirs \- 
https://bugs.debian.org/688842 Fix CPAN&#8203;::FirstTime defaults with 
nonexisting site dirs if a parent is writable
     DEBPKG&#8203;:fixes/memoize\_storable\_nstore \- \[rt\.cpan\.org \#77790\] 
https://bugs.debian.org/587650 Memoize&#8203;::Storable&#8203;: respect 'nstore' 
option not respected
     DEBPKG&#8203;:debian/regen\-skip \- Skip a regeneration check in unrelated 
git repositories
     DEBPKG&#8203;:debian/makemaker\-pasthru \- https://bugs.debian.org/758471 
Pass LD settings through to subdirectories
     DEBPKG&#8203;:debian/makemaker\-manext \- https://bugs.debian.org/247370 
Make EU&#8203;::MakeMaker honour MANnEXT settings in generated manpage headers
     DEBPKG&#8203;:debian/devel\-ppport\-reproducibility \- 
https://bugs.debian.org/801523 Sort the list of XS code files when 
generating RealPPPort\.xs
     DEBPKG&#8203;:debian/encode\-unicode\-bom\-doc \- 
https://bugs.debian.org/798727 Document Debian backport of 
Encode&#8203;::Unicode fix
     DEBPKG&#8203;:debian/kfreebsd\-softupdates \- https://bugs.debian.org/796798 
Work around Debian Bug\#796798
     DEBPKG&#8203;:fixes/autodie\-scope \- https://bugs.debian.org/798096 Fix a 
scoping issue with "no autodie" and the "system" sub
     DEBPKG&#8203;:fixes/crosscompile\-no\-targethost \- \[23695c0\] \[perl \#127234\] 
Fix the Configure escape with usecrosscompile but no targethost
     DEBPKG&#8203;:fixes/memoize\-pod \- \[rt\.cpan\.org \#89441\] Fix POD errors in 
Memoize
     DEBPKG&#8203;:fixes/ok\-pod \- Added encoding for pod\.
     DEBPKG&#8203;:debian/hurd\-softupdates \- https://bugs.debian.org/822735 Fix 
t/op/stat\.t failures on hurd
     DEBPKG&#8203;:fixes/nntp\_docs \- https://bugs.debian.org/51962 Net&#8203;::NNTP&#8203;: 
Correct innd/nnrpd confusion in relation to Reader option
     DEBPKG&#8203;:fixes/math\_complex\_doc\_great\_circle \- 
https://bugs.debian.org/697567 \[rt\.cpan\.org \#114104\] Math&#8203;::Trig&#8203;: clarify 
definition of great\_circle\_midpoint
     DEBPKG&#8203;:fixes/math\_complex\_doc\_see\_also \- 
https://bugs.debian.org/697568 \[rt\.cpan\.org \#114105\] Math&#8203;::Trig&#8203;: add 
missing SEE ALSO
     DEBPKG&#8203;:fixes/math\_complex\_doc\_angle\_units \- 
https://bugs.debian.org/731505 \[rt\.cpan\.org \#114106\] Math&#8203;::Trig&#8203;: 
document angle units
     DEBPKG&#8203;:fixes/cpan\_web\_link \- https://bugs.debian.org/367291 CPAN&#8203;: 
Add link to main CPAN web site
     DEBPKG&#8203;:fixes/time\_piece\_doc \- https://bugs.debian.org/817925 
Time&#8203;::Piece&#8203;: Improve documentation for add\_months and add\_years
     DEBPKG&#8203;:fixes/perlbug\-refactor \- https://bugs.debian.org/822463 
\[perl \#128020\] perlbug&#8203;: Refactor duplicated file reading code
     DEBPKG&#8203;:fixes/perlbug\-linewrap \- https://bugs.debian.org/822463 
\[perl \#128020\] perlbug&#8203;: wrap overly long lines
     DEBPKG&#8203;:fixes/hurd\_sigaction \- https://bugs.debian.org/825016 
\[d54f4ed\] ext/POSIX/t/sigaction\.t&#8203;: Skip uid and pid tests on GNU/Hurd
     DEBPKG&#8203;:fixes/hurd\_hints \- \[4694301\] https://bugs.debian.org/825020 
\[perl \#128279\] Modify hints for Hurd per Debian ticket 825020\.
     DEBPKG&#8203;:fixes/extutils\-parsexs\-reproducibility \- \[perl \#128517\] 
https://bugs.debian.org/829296 Make the output of ExtUtils&#8203;::ParseXS 
reproducible
     DEBPKG&#8203;:debian/CVE\-2016\-1238/sitecustomize\-in\-etc \- Look for 
sitecustomize\.pl in /etc/perl rather than sitelib on Debian systems
     DEBPKG&#8203;:debian/CVE\-2016\-1238/test\-suite\-without\-dot \- \[perl \#127810\] 
Patch unit tests to explicitly insert "\." into @&#8203;INC when needed\.
     DEBPKG&#8203;:debian/CVE\-2016\-1238/eumm\-without\-dot \- \[perl \#127810\] Add 
PERL\_USE\_UNSAFE\_INC support to EU&#8203;::MM for fortify\_inc support\.
     DEBPKG&#8203;:debian/CVE\-2016\-1238/cpan\-without\-dot \- \[perl \#127810\] Set 
PERL\_USE\_UNSAFE\_INC for cpan usage
     DEBPKG&#8203;:debian/document\_inc\_removal \- Document in perlvar that we 
remove '\.' from @&#8203;INC by default
     DEBPKG&#8203;:fixes/extutils\_makemaker\_reproducible \- 
https&#8203;://bugs\.debian\.org/835815 https://bugs.debian.org/834190 Make 
perllocal\.pod files reproducible
     DEBPKG&#8203;:debian/CVE\-2016\-1238/remove\-inc\-test \- Remove test for '\.' 
in @&#8203;INC as it might not be
     DEBPKG&#8203;:debian/customized \- Update customized\.dat for files patched 
in Debian
     DEBPKG&#8203;:fixes/file\_path\_hurd\_errno \- File\-Path&#8203;: Fix test failure in 
Hurd due to hard\-coded ENOENT
     DEBPKG&#8203;:debian/hppa\_op\_optimize\_workaround \- 
https://bugs.debian.org/838613 Temporarily lower the optimization of 
op\.c on hppa due to gcc\-6 problems
     DEBPKG&#8203;:fixes/test\-builder\-warning \- https://bugs.debian.org/840968 
Silence a 'used only once' warning in Test&#8203;::Builder
     DEBPKG&#8203;:fixes/longdblinf\-randomness \- \[dd68853\] \[perl \#130133\] 
https://bugs.debian.org/844752 Configure&#8203;: fix garbage filtering with 
80\-bit long doubles
     DEBPKG&#8203;:debian/installman\-utf8 \- https://bugs.debian.org/840211 
Generate man pages with UTF\-8 characters
     DEBPKG&#8203;:fixes/list\_assign\_leak \- \[1050723\] \[perl \#130766\] 
https://bugs.debian.org/855064 avoid a leak in list assign from/to magic 
values
     DEBPKG&#8203;:fixes/perlfunc\_inc\_doc \- \[a03e9f8\] 
https://bugs.debian.org/839536 \[perl \#130832\] Documentation fixes for 
'\.' possibly no longer being in @&#8203;INC
DEBPKG&#8203;:fixes/Compress\-Raw\-Zlib\-2\.071\-Adapt\-tests\-to\-zlib\-1\.2\.11\.patch \- 
\[PATCH\] Adapt tests to zlib 1\.2\.11
     DEBPKG&#8203;:fixes/0001\-Adapt\-tests\-to\-zlib\-1\.2\.11\.patch \- \[PATCH\] Adapt 
tests to zlib 1\.2\.11



---
@&#8203;INC for perl 5\.24\.1&#8203;:
/home/jima/perl5/lib/perl5/5\.24\.1/x86\_64\-linux\-gnu\-thread\-multi
     /home/jima/perl5/lib/perl5/5\.24\.1
/home/jima/perl5/lib/perl5/x86\_64\-linux\-gnu\-thread\-multi
     /home/jima/perl5/lib/perl5
     /home/jima/lib/perl
     /etc/perl
     /usr/local/lib/x86\_64\-linux\-gnu/perl/5\.24\.1
     /usr/local/share/perl/5\.24\.1
     /usr/lib/x86\_64\-linux\-gnu/perl5/5\.24
     /usr/share/perl5
     /usr/lib/x86\_64\-linux\-gnu/perl/5\.24
     /usr/share/perl/5\.24
     /usr/local/lib/site\_perl
     /usr/lib/x86\_64\-linux\-gnu/perl\-base



---
Environment for perl 5\.24\.1&#8203;:
     HOME=/home/jima
     LANG=en\_US\.UTF\-8
     LANGUAGE=en\_US
     LC\_COLLATE=C
     LD\_LIBRARY\_PATH \(unset\)
     LOGDIR \(unset\)
PATH=/home/jima/perl5/bin&#8203;:/home/jima/bin&#8203;:/home/jima/jima\_tools/x86\_64/bin&#8203;:/home/jima/jima\_tools/bin&#8203;:/usr/bin&#8203;:/bin&#8203;:/usr/sbin&#8203;:/sbin&#8203;:/usr/bin/X11&#8203;:/usr/local/bin&#8203;:/usr/local/sbin&#8203;:/usr/games&#8203;:/usr/local/games&#8203;:/usr/lib/jvm/java\-8\-oracle/bin&#8203;:/usr/lib/jvm/java\-8\-oracle/db/bin&#8203;:/usr/lib/jvm/java\-8\-oracle/jre/bin&#8203;:\.
     PERL5LIB=/home/jima/perl5/lib/perl5&#8203;:/home/jima/lib/perl
     PERL\_BADLANG \(unset\)
     PERL\_LOCAL\_LIB\_ROOT=/home/jima/perl5
     PERL\_MB\_OPT=\-\-install\_base "/home/jima/perl5"
     PERL\_MM\_OPT=INSTALL\_BASE=/home/jima/perl5
     SHELL=/bin/bash

**Comment by p5pRT** [comment]
### From zefram@fysh.org

>      print $outfh \<$infh>       \#Syntax Error

"$outfh \<$infh" parses as a comparison expression\.  A syntax error
of this kind is to be expected from such ambiguity; it is not a bug\.
You can avoid this by wrapping $outfh in braces\.

\-zefram

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @jimav

Ah\, thanks\.    I had no idea that braces were allowed to be used like that\.
Can you briefly explain why this is legal syntax?   Is it a special case 
just for indirect object syntax

    methodname \{expr yielding object ref\} args\.\.\.

or is this an example of a more general concept?   Normally I would 
expect \{\.\.\.\} where a value is expected to be a hash constructor 
\(obviously not the case here\)\.   So I'm surprised\.

Just point me to the docs and I'll read\.

Thanks again\,
\-Jim


On 11/11/2017 04&#8203;:16 PM\, Zefram via RT wrote&#8203;:
>>       print $outfh \<$infh>       \#Syntax Error
> "$outfh \<$infh" parses as a comparison expression\.  A syntax error
> of this kind is to be expected from such ambiguity; it is not a bug\.
> You can avoid this by wrapping $outfh in braces\.
>
> \-zefram
>

**Comment by p5pRT** [comment]
### From zefram@fysh.org

Jim Avera wrote&#8203;:
>Can you briefly explain why this is legal syntax?

It's legal by fiat\.  A braced block is one of the permitted forms of
indirect object\.  The main purpose of it is to permit an arbitrarily
complicated subexpression to be used as an indirect object\.  It's not
well documented\, but the entry for "print" in perlfunc\(1\) describes it\,
with examples\.

>                                                    Normally I would expect
>\{\.\.\.\} where a value is expected to be a hash constructor \(obviously not the
>case here\)\.

There's an ambiguity between hash constructors and code blocks in many
situations\, including here\.  perl guesses which is meant\, and occasionally
gets it wrong\.  You can force the correct interpretation by opening the
construct with "\+\{" for a hash constructor or "\{;" for a code block\.

\-zefram

**Comment by p5pRT** [comment]
### From @xsawyerx



On 11/12/2017 09&#8203;:20 AM\, Jim Avera wrote&#8203;:
> Ah\, thanks\.    I had no idea that braces were allowed to be used like
> that\.
> Can you briefly explain why this is legal syntax?   Is it a special
> case just for indirect object syntax
>
>    methodname \{expr yielding object ref\} args\.\.\.
>
> or is this an example of a more general concept?   Normally I would
> expect \{\.\.\.\} where a value is expected to be a hash constructor
> \(obviously not the case here\)\.   So I'm surprised\.
>
> Just point me to the docs and I'll read\.

Hi\, Jim\. :\)

The docs for print \(perldoc \-f print\) include the following paragraph&#8203;:

        If you're storing handles in an array or hash\, or in general
whenever
        you're using any expression more complex than a bareword handle or a
        plain\, unsubscripted scalar variable to retrieve it\, you will
have to
        use a block returning the filehandle value instead\, in which
case the
        LIST may not be omitted&#8203;:

            print \{ $files\[$i\] \} "stuff\\n";
            print \{ $OK ? \*STDOUT : \*STDERR \} "stuff\\n";

This is what includes the ability to do \(as also mentioned by Perl Best
Practices\) the following&#8203;:

&nbsp;    print \{$fh\} "foo";

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'rejected'

**Comment by p5pRT** [comment]
### From @jimav

Thank you Zefram & Sawyer for the explanation\.  I appreciate it\.

Sorry for the bogus bugrep
\-Jim

**Comment by p5pRT** [comment]
### From @xsawyerx



On 11/13/2017 08&#8203;:07 AM\, Jim Avera wrote&#8203;:
> Thank you Zefram & Sawyer for the explanation\.  I appreciate it\.
>
> Sorry for the bogus bugrep

Don't worry about it\. :\)
